napari/config/_tests/test_config.py

```diff
@@ -0,0 +1,549 @@
+        "NAPARI_G": "/not/parsable/as/literal",
+        "FOO": "not included",
+    }
+
+    expected = {
+        "a_b": 123,
+        "c": True,
+        "d": "hello",
+        "e": {"x": 123, "y": 456},
+        "f": [1, 2, "3"],
+        "g": "/not/parsable/as/literal",
+    }
+
+    res = config.core.collect_env(env)
+    res.pop("_dirty")
+    assert res == expected
+
+
+def test_collect(tmp_path):
+    a = {"x": 1, "y": {"a": 1}}
+    b = {"x": 2, "z": 3, "y": {"b": 2}}
+    env = {"NAPARI_W": 4}
+
+    expected = {"w": 4, "x": 2, "y": {"a": 1, "b": 2}, "z": 3}
+
+    fn1 = tmp_path / "fn1.yaml"
+    fn2 = tmp_path / "fn2.yaml"
+
+    with open(fn1, "w") as f:
+        yaml.dump(a, f)
+    with open(fn2, "w") as f:
+        yaml.dump(b, f)
+
+    conf = config.collect([fn1, fn2], env=env)
+    conf.pop("_dirty", None)
+    assert conf == expected
+
+
+def test_collect_env_none():
+    os.environ["NAPARI_FOO"] = "bar"
+    try:
+        conf = config.collect([])
+        conf.pop("_dirty", None)
+        assert conf == {"foo": "bar"}
+    finally:
+        del os.environ["NAPARI_FOO"]
+
+
+def test_get():
+    d = {"x": 1, "y": {"a": 2}}
+
+    assert config.get("x", config=d) == 1
+    assert config.get("y.a", config=d) == 2
+    assert config.get("y.b", 123, config=d) == 123
+    with pytest.raises(KeyError):
+        config.get("y.b", config=d)
+
+
+def test_pop():
+    d = {"x": 1, "y": {"a": 2, "b": 3}}
+
+    assert config.pop("y.a", config=d) == 2
+    d.pop("_dirty", None)
+    assert d == {"x": 1, "y": {"b": 3}}
+    assert config.pop("x", config=d) == 1
+    d.pop("_dirty", None)
+    assert d == {"y": {"b": 3}}
+    assert config.pop("y.c", 123, config=d) == 123
+    assert config.pop("y", config=d) == {"b": 3}
+    d.pop("_dirty", None)
+    assert d == {}
+
+    with pytest.raises(KeyError):
+        config.pop("z", config=d)
+
+
+def test_ensure_file(tmpdir):
+    a = {"x": 1, "y": {"a": 1}}
+    b = {"x": 123}
+
+    source = os.path.join(str(tmpdir), "source.yaml")
+    dest = os.path.join(str(tmpdir), "dest")
+    destination = os.path.join(dest, "source.yaml")
+
+    with open(source, "w") as f:
+        yaml.dump(a, f)
+
+    config.ensure_file(source=source, destination=dest, comment=False)
+
+    with open(destination) as f:
+        result = yaml.safe_load(f)
+    assert result == a
+
+    # don't overwrite old config files
+    with open(source, "w") as f:
+        yaml.dump(b, f)
+
+    config.ensure_file(source=source, destination=dest, comment=False)
+
+    with open(destination) as f:
+        result = yaml.safe_load(f)
+    assert result == a
+
+    os.remove(destination)
+
+    # Write again, now with comments
+    config.ensure_file(source=source, destination=dest, comment=True)
+
+    with open(destination) as f:
+        text = f.read()
+    assert "123" in text
+
+    with open(destination) as f:
+        result = yaml.safe_load(f)
+    assert not result
+
+
+def test_set():
```

Comment:
When I do

```python
config.set(main_window.theme='dark')
```

in the console I get this

```python-traceback
config.set(main_window.theme='dark')
  File "<ipython-input-7-95730f83dd74>", line 1
    config.set(main_window.theme='dark')
              ^
SyntaxError: keyword can't be an expression
```
similarly for
```python
config.set('main_window.theme'='dark')
```
but 
```python
config.set({'main_window.theme':'dark'})
```
works just fine. Not sure if this is expected, or I am doing something wrong

Reply:
Isn't this just python? Keyword argument names have the same constraints as variable names, i.e. unambiguously not expressions. They can't include dots, because that resolves to `pre_dot.__getattr__(post_dot)`. Putting quotes around it would be interpreted as a string literal, also not a valid variable name. The last one, being a dict literal from str to str, is fine. It looks like hierarchical keys in keyword arguments can be separated with a double underscore instead (as is done later in the tests): `config.set(main_window__theme="dark")`.

Reply:
Yes, @clbarnes has it right. @sofroniewn, if you want  to use the dot notation, it must be a `dict` key, as you found.  Otherwise you have to use the double underscore notation for nested keys